File: others/2400-2500/2407. 最长递增子序列 II.cpp
//
// Created by 123 on 25-6-12.
//
#include "bits/stdc++.h"
using namespace std;
using ll = long long;
constexpr int MOD = 1'000'000'007;
constexpr int DIR[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

class Solution {
public:
	int lengthOfLIS(vector<int> &nums, int k) {
		int m = ranges::max(nums);
		vector<int> mx(4 * m);

		auto update = [&](this auto&&update, int o, int l, int r, int target, int val)-> void {
			if (l == r) {
				mx[o] = val;
				return;
			}

			int mid = (l + r) / 2;
			if (target <= mid) {
				update(o * 2, l, mid, target, val);
			} else {
				update(o * 2 + 1, mid + 1, r, target, val);
			}
			mx[o] = max(mx[o * 2], mx[o * 2 + 1]);
		};

		auto query_mx = [&](this auto&&query_mx,int o, int l, int r, int L, int R) -> int {
			if (L <= l && r <= R) return mx[o];

			int mid = (l + r) / 2;
			int res = 0;
			if (L <= mid) {
				res = max(res, query_mx(o * 2, l, mid, L, R));
			}
			if (R > mid) {
				res = max(res, query_mx(o * 2 + 1, mid + 1, r, L, R));
			}
			return res;
		};

		for (int x: nums) {
			if (x == 1) {
				update(1, 1, m, 1, 1);
			} else {
				int t = 1 + query_mx(1, 1, m, max(x - k, 1), x - 1);
				update(1, 1, m, x, t);
			}
		}
		return mx[1];
	}
};

File: leetcode/2400-2500/2407. 最长递增子序列 II.py
from typing import List


class Solution:
    def lengthOfLIS(self, nums: List[int], k: int) -> int:
        u = max(nums)
        mx = [0] * 4 * u

        def update(o: int, l: int, r: int, target: int, val: int) -> int:
            if l == r:
                mx[o] = val
                return
            mid = (l + r) // 2
            if target <= mid:
                update(o * 2, l, mid, target, val)
            else:
                update(o * 2 + 1, mid + 1, r, target, val)
            mx[o] = max(mx[o * 2], mx[o * 2 + 1])

        def query(o: int, l: int, r: int, L: int, R: int) -> int:
            if L <= l and R >= r:
                return mx[o]
            res = 0
            mid = (l + r) // 2
            if L <= mid:
                res = max(res, query(o * 2, l, mid, L, R))
            if R > mid:
                res = max(res, query(o * 2 + 1, mid + 1, r, L, R))
            return res

        for x in nums:
            if x == 1:
                update(1, 1, u, 1, 1)
            else:
                res = 1 + query(1, 1, u, max(1, x - k), x - 1)
                update(1, 1, u, x, res)
        return mx[1]


class Solution:
    def lengthOfLIS(self, nums: List[int], k: int) -> int:
        m = max(nums)
        mx = [0] * 4 * m

        def update(o, l, r, target, val):
            if l == r:
                mx[o] = val
                return
            mid = (l + r) // 2
            if target <= mid:
                update(o * 2, l, mid, target, val)
            else:
                update(o * 2 + 1, mid + 1, r, target, val)
            mx[o] = max(mx[o * 2], mx[o * 2 + 1])

        def query(o, l, r, L, R):
            if L <= l and R >= r:
                return mx[o]
            mid = (l + r) // 2
            res = 0
            if L <= mid:
                res = max(res, query(o * 2, l, mid, L, R))
            if R > mid:
                res = max(res, query(o * 2 + 1, mid + 1, r, L, R))
            return res

        for x in nums:
            if x == 1:
                update(1, 1, m, 1, 1)
            else:
                res = query(1, 1, m, max(x - k, 1), x - 1) + 1
                update(1, 1, m, x, res)
        return mx[1]
